Rewrite `cirq.stratified_circuit` following new Transformer API and primitives.

tanujkhattar opened the pull request:
- Moves `cirq/optimizers/stratify.py` to `cirq/transformers/stratify.py`. 
- Part of Organization (and deprecation) of cirq-core/optimizers in cirq-core/transformers https://github.com/quantumlib/Cirq/issues/4722
- Follows the new Transformer API Compiling: Circuit Transformers API https://github.com/quantumlib/Cirq/issues/4483
- Supports no compile tags NoCompile Tag for optimizers NoCompile Tag for optimizers https://github.com/quantumlib/Cirq/issues/4253[](https://github.com/tanujkhattar)

MichaelBroughton commented on cirq-core/cirq/transformers/stratify.py:
```diff
@@ -48,8 +64,12 @@ def stratified_circuit(
     with operations from the same category. There is no guarantee that the
     resulting circuit will be optimally packed under this constraint.
```
The docs description here seems super long winded, can we tighten this up ( maybe just cut it to the last paragraph or say something to the effect: "This transforms the given circuit to ensure that no operations of different categories are found in the same moment. Makes no optimality guarantees" ?

tanujkhattar replied:
done

MichaelBroughton commented on cirq-core/cirq/transformers/stratify.py:
```diff
@@ -73,58 +93,87 @@ def stratified_circuit(
 
 
-def stratify_circuit(classifiers: Iterable[Classifier], circuit: circuits.Circuit):
+def stratify_circuit(
```
Should we make this a private method with a leading underscore just to make clear we don't expect people to use it ?

tanujkhattar replied:
done

MichaelBroughton reviewed (commented):
Looks great. Took a quick look around and this transformer doesn't seem to get heavy use in the library (it also has exponential runtime in the number of categories).

Should we just cut it outright ? If not, I've left a few comments and we should be good to merge.

MichaelBroughton commented on cirq-core/cirq/transformers/stratify.py:
```diff
@@ -73,58 +93,87 @@ def stratified_circuit(
+    assert len(stratified_circuit) == len(circuit) * num_categories
+
+    # Try to move operations to the left to reduce circuit depth, preserving stratification.
+    for curr_idx, moment in enumerate(stratified_circuit):
+        curr_category = curr_idx % num_categories
+        if curr_category == 0:
+            # Moment containing tagged operations to be ignored.
+            continue
+        batch_removals: List[Tuple[int, 'cirq.Operation']] = []
+        batch_inserts: List[Tuple[int, 'cirq.Operation']] = []
+        for op in moment:
+            prv_idx = stratified_circuit.prev_moment_operating_on(op.qubits, curr_idx)
+            prv_idx = 0 if prv_idx is None else prv_idx + 1
+            prv_category = prv_idx % num_categories
+            should_move_to_next_batch = curr_category < prv_category
+            prv_idx += curr_category - prv_category + num_categories * should_move_to_next_batch
+            assert prv_idx <= curr_idx and prv_idx % num_categories == curr_idx % num_categories
+            if prv_idx < curr_idx:
+                batch_inserts.append((prv_idx, op))
+                batch_removals.append((curr_idx, op))
+        stratified_circuit.batch_remove(batch_removals)
+        stratified_circuit.batch_insert_into(batch_inserts)
```
Ahhh right, so it's never the case that you might need an extra category batch at the end to handle overflow for something like:

```python3
circuit = cirq.Circuit(cirq.X(q1), cirq.CX(q2, q3), cirq.CCX(q4,q5,q6), cirq.X(q6), cirq.CX(q5,q4), cirq.CCX(q3,q2,q1))
```
Where you want to stratify all the two one, two and three qubit gates separately
```
In [16]: cirq.stratified_circuit(c2, categories=[lambda op: len(op.qubits) == 1,
    ...:  lambda op: len(op.qubits)==2, lambda op: len(op.qubits) == 3])
Out[16]: 
(0, 0): ───────X───────X───
                       │
(0, 1): ───────────@───@───
                   │   │
(0, 2): ───────────X───@───

(0, 3): ───@───────X───────
           │       │
(0, 4): ───@───────@───────
           │
(0, 5): ───X───X───────────

```
Output is the same for the original implementation and yours:
```
(0, 0): ───────X───────X───
                       │
(0, 1): ───────────@───@───
                   │   │
(0, 2): ───────────X───@───

(0, 3): ───@───────X───────
           │       │
(0, 4): ───@───────@───────
           │
(0, 5): ───X───X───────────


```
Weird. In my head I was convinced there could be spillover, but we're good now.

tanujkhattar commented on cirq-core/cirq/transformers/stratify.py:
```diff
@@ -48,8 +64,12 @@ def stratified_circuit(
     Args:
         circuit: The circuit whose operations should be re-arranged.
+        context: `cirq.TransformerContext` storing common configurable options for transformers.
         categories: A list of classifiers picking out certain operations.
```
We already describe all the valid types in the description of this arg? 

```
There are several ways to specify a classifier. You can pass in a gate instance 
(e.g. `cirq.X`), a gate type (e.g. `cirq.XPowGate`), an operation instance 
(e.g. `cirq.X(cirq.LineQubit(0))`), an operation type (e.g.`cirq.CircuitOperation`), 
or an arbitrary operation predicate (e.g. `lambda op: len(op.qubits) == 2`).
```

I think it's useful to allow these types instead of directly requiring lambda's because an iterable of lambda's would be messy to write. For example:

```python
>>> categories = [cirq.X, cirq.Y, cirq.Z] # Way-1: Much cleaner 
>>> categories = [lambda op: isinstance(op.gate, cirq.X), lambda op: isinstance(op.gate, cirq.Y), lambda op: isinstance(op.gate, cirq.Z)] # Way-2: Much worse 
```

tanujkhattar commented on cirq-core/cirq/transformers/stratify_test.py:
```diff
@@ -228,7 +251,9 @@ def test_greedy_merging():
+        cirq.stratified_circuit(input_circuit, categories=[cirq.X]), expected
+    )
 
 
```
CCOs are indeed tricky! Fixed a bug, added a test and filed https://github.com/quantumlib/Cirq/issues/4976 -- which I'm sure is going to be a cause of bugs with CCOs at a lot more places in cirq :))

tanujkhattar commented:
> Should we just cut it outright ? If not, I've left a few comments and we should be good to merge.

The number of categories in practice are usually pretty small, for example: align all 1 and 2q gates in separate moments. Therefore, I think this is a useful transformer to keep. 

I've addressed your comments. PTAL.

MichaelBroughton reviewed (approved):
LGTM

MichaelBroughton commented:
Could you try pushing a no-op to the branch here to retrigger the CI ? I can't seem to get it to go.....
